maestrowf/maestro.py

```diff
@@ -72,20 +98,146 @@ def status_study(args):
+                              'Number Restarts']
+
+                for row in range(num_rows):
+                    step_table = Table(
+                        box=box.SIMPLE_HEAVY,
+                        show_header=False,
+                    )
+                    step_table.add_column("key")
+                    step_table.add_column("val")
+
+                    step_table.add_row("STEP:",
+                                       status['Step Name'][row],
+                                       style='Step Name')
+                    step_table.add_row("WORKSPACE:",
+                                       status['Workspace'][row],
+                                       style='Workspace')
+
+                    step_table.add_row("", "")
+                    step_details = Table.grid(padding=1)
+                    step_details.add_column("details")
+
+                    step_info = Table(title="Step Details",
+                                      # show_edge=True,
+                                      show_header=False,
+                                      show_lines=True,
+                                      # show_footer=True,
+                                      box=box.HORIZONTALS)
+
+                    step_info.add_column("key")
+                    step_info.add_column("val")
+                    for nom_row_cnt, detail_row in enumerate(detail_rows):
+                        if detail_row == 'State':
+                            row_style = 'State'
+                        else:
+                            if nom_row_cnt % 2 == 0:
+                                row_style = 'row_style'
+                            else:
+                                row_style = 'row_style'
+
+                        step_info.add_row(detail_row,
+                                          status[detail_row][row],
+                                          style=row_style)
+
+                    step_details.add_column("scheduler")
+                    step_sched = Table(title="Scheduler Details",
+                                       show_header=False,
+                                       show_lines=True,
+                                       box=box.HORIZONTALS)
+                    step_sched.add_column("key")
+                    step_sched.add_column("val")
+                    for nom_row_cnt, sched_row in enumerate(sched_rows):
+                        if nom_row_cnt % 2 == 0:
+                            row_style = 'row_style'
+                        else:
+                            row_style = 'row_style'
+
+                        step_sched.add_row(sched_row,
+                                           status[sched_row][row],
+                                           style=row_style)
+
+                    step_details.add_row(step_info, step_sched)
+
+                    step_table.add_row('', step_details)
+                    stat_table.add_row(step_table, end_section=True)
```

Comment:
It might be worth having a factory for the different table functions.

Reply:
Actually, I think I did this somewhere in one of the command line args.. you can have the factory return the valid choices for the `choices=` parameter in `add_argument`. Then when any new options are added, it automatically gets updated.

Reply:
oh, that sounds much nicer.  layout factory located elsewhere it is

Reply:
What I normally do is put the factory classes in the `__init__` files for the module "above" the items being controlled, but that's really useful for things like adapters and other interface items. In this case, since it's part of the front end of Maestro, I could see it being put in `maestro.py` or in the top level `__init__` file for general use. To me this is UI specific to Maestro.

Reply:
alright, sounds good; just wasn't sure if you wanted to clutter up the main file.  i'll stick it in there

Reply:
Ok, given the filtering tools are gonna need to live in utils, think the concrete renderers and their registration is going to need to move into maestro.py so the filters can actually be imported.